contracts/external/cw-payroll/src/contract.rs

```diff
@@ -0,0 +1,589 @@
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    cw_ownable::initialize_owner(deps.storage, deps.api, msg.owner.as_deref())?;
+
+    // Check all params and create vesting payment with status based on denom
+    let vp = match msg.params.denom {
+        UncheckedDenom::Native(_) => {
+            // Validate all vesting payment params
+            let checked_params = msg.params.into_checked(deps.as_ref())?;
+
+            // Check amount sent matches amount in vesting payment
+            if checked_params.amount != must_pay(&info, &checked_params.denom.to_string())? {
+                return Err(ContractError::AmountDoesNotMatch);
+            }
+
+            // Create a new vesting payment
+            VestingPayment::new(checked_params)?
+        }
+        UncheckedDenom::Cw20(_) => {
+            // If configured for cw20, check no native funds included
+            nonpayable(&info)?;
+
+            // Validate all vesting payment params
+            let checked_params = msg.params.into_checked(deps.as_ref())?;
+
+            // Create new vesting payment
+            let mut vesting_payment = VestingPayment::new(checked_params)?;
+
+            // Set status to unfunded
+            vesting_payment.status = VestingPaymentStatus::Unfunded;
+
+            vesting_payment
+        }
+    };
+
+    VESTING_PAYMENT.save(deps.storage, &vp)?;
+
+    Ok(Response::new()
+        .add_attribute("method", "instantiate")
+        .add_attribute("owner", msg.owner.unwrap_or_else(|| "None".to_string())))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Receive(msg) => execute_receive_cw20(env, deps, info, msg),
+        ExecuteMsg::Distribute {} => execute_distribute(env, deps),
+        ExecuteMsg::Cancel {} => execute_cancel_vesting_payment(env, deps, info),
+        ExecuteMsg::UpdateOwnership(action) => execute_update_owner(deps, info, env, action),
+        ExecuteMsg::Delegate { validator, amount } => {
+            execute_delegate(env, deps, info, validator, amount)
+        }
+        ExecuteMsg::Redelegate {
+            src_validator,
+            dst_validator,
+            amount,
+        } => execute_redelegate(env, deps, info, src_validator, dst_validator, amount),
+        ExecuteMsg::Undelegate { validator, amount } => {
+            execute_undelegate(env, deps, info, validator, amount)
+        }
+        ExecuteMsg::WithdrawDelegatorReward { validator } => {
+            execute_withdraw_rewards(env, deps, validator)
+        }
+    }
+}
+
+pub fn execute_receive_cw20(
+    _env: Env,
+    deps: DepsMut,
+    info: MessageInfo,
+    receive_msg: Cw20ReceiveMsg,
+) -> Result<Response, ContractError> {
+    // Only accepts cw20 tokens
+    nonpayable(&info)?;
```

Comment:
i'm imagining a scenario in which the payer does not have all of the funds available for the entire vesting contract upfront. Are they to keep creating new vesting contracts on an ongoing basis in this case? that might cause a large number of contract instantiations (which I'm not sure is a concern in and of itself). What do you think about the ability to re-fund and extend an existing vesting payment? unsure if this complicates the use of curves as well

Reply:
Hmmmmm, could be a nice v2. Actually took out pausing as well because I felt it wasn't essential for the MVP.

Instantiating contracts is cheap.

I think the appeal of the vesting contract from the `recipients` perspective is that they no all the funds for the payment are in the contract. There is a nice piece of mind about that.

Reply:
V2 could support pausing or updating the existing payment.

Unless we can think of a clean way to do it, I think it complicates things a bit. Can always cancel and instantiate a new contract for now.